contracts/src/main/kotlin/com/gevamu/contracts/PaymentContract.kt

```diff
@@ -16,21 +16,193 @@
+        return tx.inputs.map { it.state.data }
+            .filterIsInstance<Payment>()
+            .first { it.uniquePaymentId == payment.uniquePaymentId }
+    }
+
+    class CreateVerifier(
+        command: CommandWithParties<CommandData>,
+        outputPayment: Payment
+    ) : PayerSignerVerifier(command, outputPayment) {
+        override fun verify() {
+            super.verify()
+            if (outputPayment.status != Payment.PaymentStatus.CREATED) {
+                throw IllegalStateException("Unexpected payment status: [${outputPayment.status}]")
+            }
+        }
+    }
+
+    class SendToGatewayVerifier(
+        command: CommandWithParties<CommandData>,
+        private val inputPayment: Payment,
+        outputPayment: Payment
+    ) : GatewaySignerVerifier(command, outputPayment) {
+        override fun verify() {
+            super.verify()
+            if (outputPayment.status != Payment.PaymentStatus.SENT_TO_GATEWAY) {
+                throw IllegalStateException("Unexpected payment status: [${outputPayment.status}]")
+            }
+            PaymentConsensusVerifier(
+                inputPayment = inputPayment,
+                outputPayment = outputPayment
+            ).verify()
+        }
+    }
+
+    class UpdateStatusVerifier(
+        command: CommandWithParties<CommandData>,
+        private val inputPayment: Payment,
+        outputPayment: Payment
+    ) : GatewaySignerVerifier(command, outputPayment) {
+        override fun verify() {
+            super.verify()
+            PaymentConsensusVerifier(
+                inputPayment = inputPayment,
+                outputPayment = outputPayment
+            ).verify()
+        }
+    }
+
+    abstract class GatewaySignerVerifier(
+        command: CommandWithParties<CommandData>,
+        outputPayment: Payment
+    ) : PayerSignerVerifier(command, outputPayment) {
+        override fun verify() {
+            super.verify()
+            if (!command.signers.contains(outputPayment.gateway.owningKey)) {
+                throw IllegalStateException("Unsigned by the Gateway party")
+            }
+        }
+    }
+
+    abstract class PayerSignerVerifier(
+        command: CommandWithParties<CommandData>,
+        outputPayment: Payment
+    ) : BaseVerifier(command, outputPayment) {
+        override fun verify() {
+            super.verify()
+            if (!command.signers.contains(outputPayment.payer.owningKey)) {
+                throw IllegalStateException("Unsigned by the Payer party")
+            }
+        }
+    }
+
+    abstract class BaseVerifier(
+        protected val command: CommandWithParties<CommandData>,
+        protected val outputPayment: Payment
+    ) : Verifier {
+        override fun verify() {
+            if (outputPayment.endToEndId.isBlank()) {
+                throw IllegalArgumentException("Invalid value of the field EndToEndId: [${outputPayment.endToEndId}]")
```

Comment:
You check for blank, be more specific in error: "cannot be blank" rather than generic "invalid value". Add output index (and optionally uniquePaymentId) to help identify the problematic output